## PO cycle — applying pipeline:fix to #895

CI is red on build errors across 8 packages:

```
pkg/dataplane/providers/grpc/stream.go:73:15: undefined: ErrEmptyChunkList
pkg/dataplane/providers/grpc/stream.go:82:40: undefined: ErrChunkCountMismatch
pkg/dataplane/providers/grpc/stream.go:87:76: undefined: ErrChunkSequenceGap
pkg/dataplane/providers/grpc/stream.go:97:53: undefined: ErrPayloadTooLarge
```

The dev agent USED the typed errors throughout `stream.go` but **never created the `errors.go` file** that declares them. The story's Files In Scope explicitly listed:

> `pkg/dataplane/providers/grpc/errors.go` (NEW FILE) — declare typed errors for chunk integrity failures (e.g., `ErrEmptyChunkList`, `ErrChunkCountMismatch`, `ErrChunkSequenceGap`, `ErrPayloadTooLarge`, `ErrTotalSizeMismatch`)

Plus per the story Acceptance Criteria:
> - [ ] `pkg/dataplane/providers/grpc/errors.go` created with typed chunk integrity error vars

### Fix direction

1. **Create `pkg/dataplane/providers/grpc/errors.go`** declaring all five sentinels:
```go
package grpc

import "errors"

var (
    ErrEmptyChunkList     = errors.New("dataplane: chunk list is empty")
    ErrChunkCountMismatch = errors.New("dataplane: TotalChunks does not match received chunk count")
    ErrChunkSequenceGap   = errors.New("dataplane: chunk sequence has gap or duplicate")
    ErrPayloadTooLarge    = errors.New("dataplane: assembled payload exceeds maximum size")
    ErrTotalSizeMismatch  = errors.New("dataplane: BulkChunk TotalSize does not match assembled length")
)
```

(Confirm exact error names by grepping `stream.go` — match what the dev agent already references.)

2. **Verify the rest of the story acceptance criteria** are present:
   - `chunksToConfigTransfer`, `chunksToDNATransfer`, `chunksToBulkTransfer` validate TotalChunks, contiguity, accumulated size cap (≤ 8 MB)
   - `chunksToBulkTransfer` additionally validates `TotalSize`
   - `bulkTransferToChunks` has `math.MaxInt32` overflow guard (matching `configTransferToChunks` and `dnaTransferToChunks`)
   - `chunkSize` const at `stream.go:17` exported as `ChunkSize` (per the v3 revision: kept in `stream.go`, NOT moved to `pkg/dataplane/types/`. The original v2 had a Move-to-types decision but v3 kept it in stream.go — verify which one is in the actual story body)
   - `config_handler.go` no longer has duplicate `chunkSize` const

3. **Mark PR #908 ready** (not draft) — the title still says "WIP: agent failed". Update title and `gh pr ready 908`.

4. Run `make test-complete` to confirm.

--- a/pkg/dataplane/providers/grpc/errors.go
+++ b/pkg/dataplane/providers/grpc/errors.go
@@ -1,0 +1,13 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2026 Jordan Ritz
+package grpc
+
+import "errors"
+
+var (
+	ErrEmptyChunkList     = errors.New("dataplane: chunk list is empty")
+	ErrChunkCountMismatch = errors.New("dataplane: TotalChunks does not match received chunk count")
+	ErrChunkSequenceGap   = errors.New("dataplane: chunk sequence has gap or duplicate")
+	ErrPayloadTooLarge    = errors.New("dataplane: assembled payload exceeds maximum size")
+	ErrTotalSizeMismatch  = errors.New("dataplane: BulkChunk TotalSize does not match assembled length")
+)

--- a/pkg/dataplane/providers/grpc/session_test.go
+++ b/pkg/dataplane/providers/grpc/session_test.go
@@ -273,9 +273,8 @@ func TestChunkRoundTrip_EmptyData(t *testing.T) {
 	// Marshal empty cfg → JSON has other fields but Data is empty
 	data, err := json.Marshal(cfg)
 	require.NoError(t, err)
-	// The JSON will be non-empty because the struct has other fields.
-	// configTransferToChunks will produce at least 1 chunk.
-	_ = data
+	// The JSON is non-empty because the struct has ID and Version set.
+	assert.NotEmpty(t, data)
 
 	chunks, err := configTransferToChunks(cfg)
 	require.NoError(t, err)

--- a/pkg/dataplane/providers/grpc/stream_test.go
+++ b/pkg/dataplane/providers/grpc/stream_test.go
@@ -1,0 +1,231 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2026 Jordan Ritz
+package grpc
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	transportpb "github.com/cfgis/cfgms/api/proto/transport"
+	"github.com/cfgis/cfgms/pkg/dataplane/types"
+)
+
+// TestChunksToConfigTransfer_ValidationErrors is a table-driven suite covering all
+// chunk integrity validation error paths for config chunk reassembly.
+func TestChunksToConfigTransfer_ValidationErrors(t *testing.T) {
+	tests := []struct {
+		name    string
+		chunks  []*transportpb.ConfigChunk
+		wantErr error
+	}{
+		{
+			name:    "empty_slice",
+			chunks:  []*transportpb.ConfigChunk{},
+			wantErr: ErrEmptyChunkList,
+		},
+		{
+			name: "total_chunks_mismatch",
+			// 2 chunks but TotalChunks=3 — count does not match.
+			chunks: []*transportpb.ConfigChunk{
+				{Data: []byte("a"), ChunkIndex: 0, TotalChunks: 3},
+				{Data: []byte("b"), ChunkIndex: 1, TotalChunks: 3},
+			},
+			wantErr: ErrChunkCountMismatch,
+		},
+		{
+			name: "sequence_gap",
+			// indices 0, 2 with TotalChunks=2 — index 1 is missing.
+			chunks: []*transportpb.ConfigChunk{
+				{Data: []byte("a"), ChunkIndex: 0, TotalChunks: 2},
+				{Data: []byte("c"), ChunkIndex: 2, TotalChunks: 2},
+			},
+			wantErr: ErrChunkSequenceGap,
+		},
+		{
+			name: "duplicate_sequence",
+			// Two chunks both at index 0 — the duplicate causes a gap at index 1.
+			chunks: []*transportpb.ConfigChunk{
+				{Data: []byte("a"), ChunkIndex: 0, TotalChunks: 2},
+				{Data: []byte("b"), ChunkIndex: 0, TotalChunks: 2},
+			},
+			wantErr: ErrChunkSequenceGap,
+		},
+		{
+			name: "payload_too_large",
+			// Two 5 MB chunks → 10 MB assembled payload exceeds maxRecvMsgSize (8 MB).
+			chunks: func() []*transportpb.ConfigChunk {
+				big := make([]byte, 5*1024*1024)
+				return []*transportpb.ConfigChunk{
+					{Data: big, ChunkIndex: 0, TotalChunks: 2},
+					{Data: big, ChunkIndex: 1, TotalChunks: 2},
+				}
+			}(),
+			wantErr: ErrPayloadTooLarge,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			_, err := chunksToConfigTransfer(tc.chunks)
+			require.Error(t, err)
+			assert.True(t, errors.Is(err, tc.wantErr), "want %v, got %v", tc.wantErr, err)
+		})
+	}
+}
+
+// TestChunksToDNATransfer_ValidationErrors is a table-driven suite covering all
+// chunk integrity validation error paths for DNA chunk reassembly.
+func TestChunksToDNATransfer_ValidationErrors(t *testing.T) {
+	tests := []struct {
+		name    string
+		chunks  []*transportpb.DNAChunk
+		wantErr error
+	}{
+		{
+			name:    "empty_slice",
+			chunks:  []*transportpb.DNAChunk{},
+			wantErr: ErrEmptyChunkList,
+		},
+		{
+			name: "total_chunks_mismatch",
+			chunks: []*transportpb.DNAChunk{
+				{Data: []byte("a"), ChunkIndex: 0, TotalChunks: 3, StewardId: "s1", TenantId: "t1"},
+				{Data: []byte("b"), ChunkIndex: 1, TotalChunks: 3, StewardId: "s1", TenantId: "t1"},
+			},
+			wantErr: ErrChunkCountMismatch,
+		},
+		{
+			name: "sequence_gap",
+			chunks: []*transportpb.DNAChunk{
+				{Data: []byte("a"), ChunkIndex: 0, TotalChunks: 2, StewardId: "s1", TenantId: "t1"},
+				{Data: []byte("c"), ChunkIndex: 2, TotalChunks: 2, StewardId: "s1", TenantId: "t1"},
+			},
+			wantErr: ErrChunkSequenceGap,
+		},
+		{
+			name: "duplicate_sequence",
+			chunks: []*transportpb.DNAChunk{
+				{Data: []byte("a"), ChunkIndex: 0, TotalChunks: 2, StewardId: "s1", TenantId: "t1"},
+				{Data: []byte("b"), ChunkIndex: 0, TotalChunks: 2, StewardId: "s1", TenantId: "t1"},
+			},
+			wantErr: ErrChunkSequenceGap,
+		},
+		{
+			name: "payload_too_large",
+			chunks: func() []*transportpb.DNAChunk {
+				big := make([]byte, 5*1024*1024)
+				return []*transportpb.DNAChunk{
+					{Data: big, ChunkIndex: 0, TotalChunks: 2, StewardId: "s1", TenantId: "t1"},
+					{Data: big, ChunkIndex: 1, TotalChunks: 2, StewardId: "s1", TenantId: "t1"},
+				}
+			}(),
+			wantErr: ErrPayloadTooLarge,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			_, err := chunksToDNATransfer(tc.chunks)
+			require.Error(t, err)
+			assert.True(t, errors.Is(err, tc.wantErr), "want %v, got %v", tc.wantErr, err)
+		})
+	}
+}
+
+// TestChunksToBulkTransfer_ValidationErrors is a table-driven suite covering all
+// chunk integrity validation error paths for bulk chunk reassembly.
+func TestChunksToBulkTransfer_ValidationErrors(t *testing.T) {
+	tests := []struct {
+		name    string
+		chunks  []*transportpb.BulkChunk
+		wantErr error
+	}{
+		{
+			name:    "empty_slice",
+			chunks:  []*transportpb.BulkChunk{},
+			wantErr: ErrEmptyChunkList,
+		},
+		{
+			name: "offset_gap",
+			// chunk[1] starts at offset 10, but chunk[0] ends at offset 5 — 5-byte gap.
+			chunks: []*transportpb.BulkChunk{
+				{Data: []byte("hello"), Offset: 0, TotalSize: 10},
+				{Data: []byte("world"), Offset: 10, TotalSize: 10},
+			},
+			wantErr: ErrChunkSequenceGap,
+		},
+		{
+			name: "payload_too_large",
+			// Two 5 MB chunks → 10 MB assembled payload exceeds maxRecvMsgSize (8 MB).
+			chunks: func() []*transportpb.BulkChunk {
+				big := make([]byte, 5*1024*1024)
+				return []*transportpb.BulkChunk{
+					{Data: big, Offset: 0, TotalSize: int64(10 * 1024 * 1024)},
+					{Data: big, Offset: int64(5 * 1024 * 1024), TotalSize: int64(10 * 1024 * 1024)},
+				}
+			}(),
+			wantErr: ErrPayloadTooLarge,
+		},
+		{
+			name: "duplicate_offset",
+			// Two chunks both at offset 0 — the duplicate causes a mismatch on the second iteration.
+			chunks: []*transportpb.BulkChunk{
+				{Data: []byte("hello"), Offset: 0, TotalSize: 10},
+				{Data: []byte("world"), Offset: 0, TotalSize: 10},
+			},
+			wantErr: ErrChunkSequenceGap,
+		},
+		{
+			name: "total_size_mismatch",
+			// Assembled payload is 13 bytes but TotalSize claims 999.
+			chunks: []*transportpb.BulkChunk{
+				{Data: []byte(`{"id":"test"}`), Offset: 0, TotalSize: 999},
+			},
+			wantErr: ErrTotalSizeMismatch,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			_, err := chunksToBulkTransfer(tc.chunks)
+			require.Error(t, err)
+			assert.True(t, errors.Is(err, tc.wantErr), "want %v, got %v", tc.wantErr, err)
+		})
+	}
+}
+
+// TestChunksToBulkTransfer_RoundTrip verifies that bulkTransferToChunks and
+// chunksToBulkTransfer are inverse operations for a normal-sized payload.
+func TestChunksToBulkTransfer_RoundTrip(t *testing.T) {
+	original := &types.BulkTransfer{
+		ID:        "bulk-rt-001",
+		StewardID: "steward-rt",
+		TenantID:  "tenant-rt",
+		Direction: "to_steward",
+		Type:      "file",
+	}
+	chunks, err := bulkTransferToChunks(original)
+	require.NoError(t, err)
+	require.NotEmpty(t, chunks)
+
+	got, err := chunksToBulkTransfer(chunks)
+	require.NoError(t, err)
+	assert.Equal(t, original.ID, got.ID)
+	assert.Equal(t, original.StewardID, got.StewardID)
+	assert.Equal(t, original.TenantID, got.TenantID)
+}
+
+// TestBulkTransferToChunks_SmallPayload verifies that bulkTransferToChunks produces
+// at least one chunk for a small payload and does not error.
+func TestBulkTransferToChunks_SmallPayload(t *testing.T) {
+	bulk := &types.BulkTransfer{
+		ID:   "bulk-small",
+		Data: []byte("payload well below any size limit"),
+	}
+	chunks, err := bulkTransferToChunks(bulk)
+	require.NoError(t, err)
+	assert.NotEmpty(t, chunks)
+}

